x/trigger/keeper/keeper.go

```diff
@@ -0,0 +1,44 @@
+package keeper
+
+import (
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/trigger/types"
+)
+
+const StoreKey = types.ModuleName
+
+// MessageRouter ADR 031 request type routing
+// https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-031-msg-service.md
+type MessageRouter interface {
+	Handler(msg sdk.Msg) MsgServiceHandler
+}
+
+// MsgServiceHandler defines a function type which handles Msg service message.
+type MsgServiceHandler = func(ctx sdk.Context, req sdk.Msg) (*sdk.Result, error)
+
+type Keeper struct {
+	storeKey storetypes.StoreKey
+	cdc      codec.BinaryCodec
+	router   MessageRouter
```

Comment:
Any reason to not just use a `IMsgServiceRouter` like the gov module?

Also, if the `MessageRouter` or `MsgServiceHandler` types are used as arguments or return types for any keeper functions (other than `NewKeeper`), they need to be defined in the `types` package instead of `keeper`.

The idea is that, when creating a mock keeper, you want to be able to do so without importing the `keeper` package. This helps reduce the risk of circular dependencies. Basically, as much as possible, you don't want to import any keeper packages from a different keeper package (or its unit tests).

Reply:
The gov module is actually using MessageRouter, but I can change it to IMsgServiceRouter. I agree that it's better to use an interface.

Reply:
Maybe they switched to a `MessageRouter` for `v0.47` and beyond then. For our stuff, it's still the `IMsgServiceRouter `.